File: benchmarks/database-core-batches.ts
/** Cross-binding durable batch-size benchmark — TypeScript lane. */
import { existsSync, mkdirSync, rmSync, statSync, writeFileSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import { tmpdir } from 'node:os'
import { PolyGraph } from '../src/graph.js'
import { BinaryStoreAdapter } from '../src/persistence/binary-store.js'
import { NodeFileIO } from '../src/persistence/node-file-io.js'

const OUT_DIR = join(dirname(fileURLToPath(import.meta.url)), 'results')
const DEFAULT_OUT = join(OUT_DIR, 'database-core-batches-ts.json')
const BATCHES = [1, 100, 500, 5_000]
function numericArg(flag: string, fallback: number): number {
  const index = process.argv.indexOf(flag)
  return index >= 0 && process.argv[index + 1] !== undefined ? Number(process.argv[index + 1]) : fallback
}
const COUNT = numericArg('--count', 5_000)
const SEED = numericArg('--seed', 42)
const outArg = process.argv.indexOf('--out')
const OUT = outArg >= 0 ? process.argv[outArg + 1] : DEFAULT_OUT
const NO_AUTO_COMPACT = 1_000_000_000

function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a = (a + 0x6d2b79f5) | 0
    let t = Math.imul(a ^ (a >>> 15), 1 | a)
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function fileBytes(dir: string, name: string): number {
  const path = join(dir, name)
  return existsSync(path) ? statSync(path).size : 0
}

function node(i: number, rng: () => number) {
  return { id: `n${i}`, type: ['user', 'post', 'comment'][i % 3], data: { idx: i, value: rng(), tag: `tag_${i % 50}` }, insertedAt: i, updatedAt: i }
}

async function runCase(root: string, batchSize: number) {
  const dir = join(root, `batch-${batchSize}`)
  mkdirSync(dir, { recursive: true })
  const io = new NodeFileIO(dir)
  const adapter = new BinaryStoreAdapter({ storeDir: '.', fileIO: io, compactThreshold: NO_AUTO_COMPACT })
  const graph = new PolyGraph(adapter)
  const rng = mulberry32(SEED)
  const writeStart = performance.now()
  for (let i = 0; i < COUNT; i++) {
    graph.addNode(node(i, rng))
    if ((i + 1) % batchSize === 0) await graph.flush()
  }
  if (COUNT % batchSize) await graph.flush()
  const writeMs = performance.now() - writeStart
  const mutationsBefore = (await adapter.getMutationsSince!(0n)).length
  const before = {
    walBytes: fileBytes(dir, 'wal.msgpack'),
    snapshotBytes: fileBytes(dir, 'snapshot.msgpack'),
    mutationLogBytes: fileBytes(dir, 'mutations.msgpack'),
  }
  const compactStart = performance.now()
  await graph.checkpoint()
  const compactMs = performance.now() - compactStart
  const after = {
    walBytes: fileBytes(dir, 'wal.msgpack'),
    snapshotBytes: fileBytes(dir, 'snapshot.msgpack'),
    mutationLogBytes: fileBytes(dir, 'mutations.msgpack'),
  }
  const verification = await graph.verify()
  await graph.dispose()
  return {
    batchSize, count: COUNT, seed: SEED, writeMs, writeOpsPerSec: COUNT / (writeMs / 1000),
    mutationCount: mutationsBefore, compactMs, verified: verification.ok, nodeCount: verification.nodeCount,
    before, after,
  }
}

async function main() {
  if (!Number.isInteger(COUNT) || COUNT < 1) throw new Error('--count must be a positive integer')
  const root = join(tmpdir(), `polypack-batches-ts-${process.pid}`)
  mkdirSync(root, { recursive: true })
  try {
    const result = { schemaVersion: 1, lang: 'ts', count: COUNT, seed: SEED, batchSizes: BATCHES, cases: [] as unknown[] }
    for (const batchSize of BATCHES) result.cases.push(await runCase(root, batchSize))
    mkdirSync(dirname(OUT), { recursive: true })
    writeFileSync(OUT, JSON.stringify(result, null, 2))
    console.log(JSON.stringify(result, null, 2))
    console.log(`Wrote ${OUT}`)
  } finally {
    rmSync(root, { recursive: true, force: true })
  }
}

main().catch(error => { console.error(error); process.exit(1) })

File: crates/polypack-graph/examples/database_core_batches.rs
//! Cross-binding durable batch-size benchmark — Rust lane.

use polypack_core::storage::FileStorage;
use polypack_core::{Node, StoreConfig};
use polypack_graph::{Graph, GraphConfig};
use serde_json::json;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;

const BATCHES: [usize; 4] = [1, 100, 500, 5_000];
const NO_AUTO_COMPACT: usize = 1_000_000_000;

struct Args { count: usize, seed: u32, out: PathBuf }

fn args() -> Args {
    let raw: Vec<String> = env::args().collect();
    let mut count = 5_000usize;
    let mut seed = 42u32;
    let mut out = PathBuf::from("benchmarks/results/database-core-batches-rust.json");
    let mut i = 1;
    while i < raw.len() {
        match raw[i].as_str() {
            "--count" => { count = raw[i + 1].parse().expect("--count must be positive"); i += 2; }
            "--seed" => { seed = raw[i + 1].parse().expect("--seed must be a number"); i += 2; }
            "--out" => { out = PathBuf::from(&raw[i + 1]); i += 2; }
            _ => { i += 1; }
        }
    }
    Args { count, seed, out }
}

struct Rng(u32);
impl Rng {
    fn next(&mut self) -> f64 {
        self.0 = self.0.wrapping_add(0x6d2b79f5);
        let mut t = (self.0 ^ (self.0 >> 15)).wrapping_mul(self.0 | 1);
        t = (t + (t ^ (t >> 7)).wrapping_mul(self.0 | 61)) ^ t;
        ((t ^ (t >> 14)) as f64) / 4294967296.0
    }
}

fn node(i: usize, rng: &mut Rng) -> Node {
    let mut data = serde_json::Map::new();
    data.insert("idx".into(), json!(i));
    data.insert("value".into(), json!(rng.next()));
    data.insert("tag".into(), json!(format!("tag_{}", i % 50)));
    Node { id: format!("n{i}"), node_type: ["user", "post", "comment"][i % 3].into(), data, vector: None, inserted_at: i as i64, updated_at: i as i64, revision: 0, activation: None }
}

fn bytes(dir: &Path, name: &str) -> u64 {
    fs::metadata(dir.join(name)).map(|m| m.len()).unwrap_or(0)
}

fn run_case(root: &Path, args: &Args, batch_size: usize) -> serde_json::Value {
    let dir = root.join(format!("batch-{batch_size}"));
    fs::create_dir_all(&dir).unwrap();
    let storage = FileStorage::open(&dir, false).unwrap();
    let config = StoreConfig { compact_threshold: NO_AUTO_COMPACT, ..Default::default() };
    let mut graph = Graph::open(Box::new(storage), config, GraphConfig::default()).unwrap();
    let mut rng = Rng(args.seed);
    let start = Instant::now();
    for i in 0..args.count {
        graph.add_node(node(i, &mut rng)).unwrap();
        if (i + 1) % batch_size == 0 { graph.flush().unwrap(); }
    }
    if args.count % batch_size != 0 { graph.flush().unwrap(); }
    let write_ms = start.elapsed().as_secs_f64() * 1000.0;
    let mutations = graph.mutation_log().unwrap().len();
    let before = json!({
        "walBytes": bytes(&dir, "wal.msgpack"),
        "snapshotBytes": bytes(&dir, "snapshot.msgpack"),
        "mutationLogBytes": bytes(&dir, "mutations.jsonl"),
    });
    let start = Instant::now();
    graph.checkpoint().unwrap();
    let compact_ms = start.elapsed().as_secs_f64() * 1000.0;
    let after = json!({
        "walBytes": bytes(&dir, "wal.msgpack"),
        "snapshotBytes": bytes(&dir, "snapshot.msgpack"),
        "mutationLogBytes": bytes(&dir, "mutations.jsonl"),
    });
    let verified = graph.verify().unwrap();
    graph.dispose().unwrap();
    json!({
        "batchSize": batch_size, "count": args.count, "seed": args.seed,
        "writeMs": write_ms, "writeOpsPerSec": args.count as f64 / (write_ms / 1000.0),
        "mutationCount": mutations, "compactMs": compact_ms,
        "verified": verified.ok, "nodeCount": verified.node_count,
        "before": before, "after": after,
    })
}

fn main() {
    let args = args();
    assert!(args.count > 0, "--count must be positive");
    let root = env::temp_dir().join(format!("polypack-batches-rust-{}", std::process::id()));
    fs::create_dir_all(&root).unwrap();
    let cases: Vec<_> = BATCHES.iter().map(|batch| run_case(&root, &args, *batch)).collect();
    let result = json!({ "schemaVersion": 1, "lang": "rust", "count": args.count, "seed": args.seed, "batchSizes": BATCHES, "cases": cases });
    if let Some(parent) = args.out.parent() { fs::create_dir_all(parent).unwrap(); }
    fs::write(&args.out, serde_json::to_string_pretty(&result).unwrap()).unwrap();
    fs::remove_dir_all(&root).ok();
    println!("{}", serde_json::to_string_pretty(&result).unwrap());
    println!("Wrote {}", args.out.display());
}
